cpp/src/io/parquet/decode_fixed.cu

```diff
@@ -285,13 +280,171 @@ __device__ inline void gpuDecodeFixedWidthSplitValues(
+    // nested hierarchies (lists), there is no 1:1 mapping of rows to values. So src_pos
+    // has to take into account the # of values we have to skip in the page to get to the
+    // desired logical row.  For flat hierarchies, skipped_leaf_values will always be 0.
+    int const src_pos = [&]() {
+      if constexpr (has_lists_t) { return thread_pos + skipped_leaf_values; }
+      return thread_pos;
+    }();
+
+    // lookup input string pointer & length. store length.
+    bool const in_range                       = (thread_pos < target_pos) && (dst_pos >= 0);
+    auto [thread_input_string, string_length] = [&]() {
+      // target_pos will always be properly bounded by num_rows, but dst_pos may be negative (values
+      // before first_row) in the flat hierarchy case.
+      if (!in_range) { return string_index_pair{nullptr, 0}; }
+      string_index_pair string_pair = gpuGetStringData(s, sb, src_pos);
+      int32_t* str_len_ptr          = reinterpret_cast<int32_t*>(ni.data_out) + dst_pos;
+      *str_len_ptr                  = string_pair.second;
+      return string_pair;
+    }();
+
+    // compute string offsets
+    size_t thread_string_offset, block_total_string_length;
+    scanner(scan_storage)
+      .ExclusiveSum(string_length, thread_string_offset, block_total_string_length);
+    __syncthreads();
+
+    // adjust for prior offset, get output string pointer
+    thread_string_offset += string_output_offset;
+    string_output_offset += block_total_string_length;
+    auto const thread_output_string = ni.string_out + thread_string_offset;
+
+    if constexpr (split_decode_t) {
+      if (in_range) {
+        auto const split_string_length = s->dtype_len_in;
+        auto const stream_length       = s->page.str_bytes / split_string_length;
+
+        for (int ii = 0; ii < split_string_length; ii++) {
+          thread_output_string[ii] = s->data_start[src_pos + ii * stream_length];
+        }
+      }
+    } else {
+      // With a dictionary, performance is MUCH better for cache hits (reuse dict entry)
+      // On a cache hit, we load a 128-byte L1/L2 cache line
+      // If each thread (e.g. 128) copies a string (128/line): 128*128 = 16kb PER BLOCK
+      // This likely overflows the L2 cache and results in a lot of cache misses
+      // Plus, perf is O(longest_length), so we're bottlenecked for non-uniform data
+
+      // Instead, the (T = block_size) threads cooperatively copy the T strings
+      // N strings are copied at a time, with M = T / N threads per string
+      __shared__ uint8_t const* inputs[block_size];
+      __shared__ uint8_t* outputs[block_size];
+      __shared__ int lengths[block_size];
+
+      // Save string pointers & lengths so threads can share the work on them
+      outputs[t] = thread_output_string;
+      inputs[t]  = reinterpret_cast<uint8_t const*>(thread_input_string);
+      lengths[t] = string_length;
+      __syncthreads();
```

Comment:
Mild optimization:  looks like you could move the __syncthreads() down to just before the "loop over all strings in this batch" for loop below.